include/seqan3/alignment/pairwise/detail/policy_optimum_tracker_simd.hpp

```diff
@@ -0,0 +1,201 @@
+#include <seqan3/core/simd/simd_traits.hpp>
+#include <seqan3/core/type_traits/lazy.hpp>
+#include <seqan3/range/views/zip.hpp>
+
+namespace seqan3::detail
+{
+
+/*!\brief A binary operation to update the alignment optimum for the vectorised global alignment algorithm.
+ * \ingroup pairwise_alignment
+ *
+ * \tparam simd_t The simd index type; must model seqan3::simd::simd_concept.
+ *
+ * \details
+ *
+ * Updates the current alignment optimum with the new score and the respective coordinate only if the coordinate
+ * matches the stored column and row index for the respective alignment. The respective coordinates need to be
+ * pre-computed.
+ */
+struct max_score_updater_simd_global
+{
+
+    /*!\brief Compares and updates the optimal score-coordinate pair.
+     * \tparam lhs_t The type of the left hand side; must model seqan3::pair_like [for more requirements see below].
+     * \tparam rhs_t The type of the right hand side; must model seqan3::pair_like [for more requirements see below].
+     *
+     * \param[in] optimal_score_coordinate_pair The optimal score and its associated matrix coordinate seen so far.
+     * \param[in] current_score_coordinate_pair The candidate score and coordinate to compare with.
+     *
+     * \details
+     *
+     * The pair on the left hand side must be assignable from the pair on the right hand side of this binary operation.
+     * Furthermore, this function expects that the first element of the given pair is the score and the second element
+     * the matrix coordinate. If the current coordinate matches the pre-initialised coordinates, then the respective
+     * score is stored as the maximum.
+     */
+    template <pair_like lhs_t, pair_like rhs_t>
+    //!\cond
+        requires (simd_concept<std::tuple_element_t<0, lhs_t>> &&
+                 simd_concept<std::tuple_element_t<0, rhs_t>> &&
+                 std::assignable_from<std::tuple_element_t<0, lhs_t>, std::tuple_element_t<0, rhs_t>>)
+    //!\endcond
+    void operator()(lhs_t && optimal_score_coordinate_pair,
+                    rhs_t && current_score_coordinate_pair) const
+    {
+        auto && [optimal_score, optimal_coordinate] = optimal_score_coordinate_pair;
+        auto && [current_cell_score, current_cell_coordinate] = current_score_coordinate_pair;
+
+        auto mask = (optimal_coordinate.col == current_cell_coordinate.col) &&
+                    (optimal_coordinate.row == current_cell_coordinate.row);
+        optimal_score = (mask) ? current_cell_score : optimal_score;
+    }
+};
+
+/*!\brief Implements the tracker to store the global optimum for a particular alignment computation.
+ * \ingroup pairwise_alignment
+ * \copydetails seqan3::detail::policy_optimum_tracker
+ */
+template <typename alignment_configuration_t, std::semiregular binary_update_operation_t>
+//!\cond
+    requires is_type_specialisation_of_v<alignment_configuration_t, configuration>
+//!\endcond
+class policy_optimum_tracker_simd :
```

Comment:
No, I won't do this. I need to handle various different simd operations based on different alignment modes and I am not putting this all into one base class. I need more members etc than in the scalar type, I need different initialisations and later different overloads for the respective functions and this is a correct abstraction to make the class more specialised for simd. So I guess, that is the correct abstraction to do. I don't understand why you are so eager to remove this subclass?

Reply:
Okay when make those classes completely independent of each other, I don't like the forced inheritance if they really share so little with each other.

Reply:
I really don‘t see your argument here. When putting shared functionality in a base class I have one class that adds nothing to it. So it would boil down to an alias nothing more. If I merge the classes, understanding the code and cases becomes ridiculously complicated. In a separate class I have a separate and clear documentation of what happens. And why should policies not inherit from each other? That doesn‘t make sense. Especially, since the one is a specialisation of the other.

Reply:
Look at the constructor and if you can't use the inherited constructor to initialise the base class that means you can't use the base class. This violates OOP for me, because the base class is left in a (possible) invalid state.

That's why you should use either a CRTP class, because you can require that your scalar derived class defines `reset_optimum` which can be used within the CRTP classes' constructor (would be my compromise). Or you merge them, in that case `reset_optimum` has the right meaning, because of a `constexpr if` (my preferred way). Or you make them complete separate (my least preferred), because how much code does the simd case reuse? Not much!

I don't buy your argument that it would make the code more difficult to understand, because you can always have well-defined functions that enable/disable depending on `requires is_vectorised` and this would minimise the number of constexpr. This makes it even easier to compare different methods, because they are in the same class and are local to each other. Having two classes, assumes that you need to know that a second class exists and this is so seqan2 style of scattering all closely coupled things over multiple locations. This is insane.

I had a really difficult time to review this, because I couldn't verify easily that you do the same or something similar in the scalar case and I think this sucks that we have such a disconnect here.

<!--
Alter, ich höre hier nur ausreden, aber okay lass uns es im seqan2 stil machen, kein klares design, komplett unmanagebaren legacy code von ANFANG an. Ich meine seqan3 war eine nette idee, aber bei jedem moment wo das für dich unpassend / unangenehm ist, fällt es bei dir runter und gilt auf einmal nicht für dich. Wer das liest ist doof :P -->

Reply:
> Look at the constructor and if you can't use the inherited constructor to initialise the base class that means you can't use the base class. This violates OOP for me, because the base class is left in a (possible) invalid state. 

Done. I call the base constructor. 😄 

> I don't buy your argument that it would make the code more difficult to understand, because you can always have well-defined functions that enable/disable depending on requires is_vectorised and this would minimise the number of constexpr. This makes it even easier to compare different methods, because they are in the same class and are local to each other. Having two classes, assumes that you need to know that a second class exists and this is so seqan2 style of scattering all closely coupled things over multiple locations. This is insane.

Sorry, but this opinion changes regularly within the team. I have been through enough examples where it doesn't matter what you code. It is always the other way that is more correct and usually due to personal preferences. I'm not taking me out of this. Now again, you want me to choose between three alternatives, which all have their pros and cons. None of them covers all the things that might come. So any alternative is likely to change in the future when we know more about the dependencies.
I am not saying that this is the final best cool design whatsoever. I am saying, there is quite a lot of things that have to be considered, added, modified etc. until all the functionality with simd and scalar and so on works with the wanted performance. If all these components are known and coded, then we can sit down and - hell yeah - optimise the last bit of code out of it. I'd love that. And I value your expertise in this. And you can point your finger at me if it ends up to be a solution of yours. But please let us first get a complete picture. My focus is on the performance now, so I can make the benchmarks etc.

Reply:
For the record. We talked this in person, and we both agreed that both approaches would have worked. We decided that we go with the static-runtime polymorphism for now, because it would be too much trouble to change it now. But we both agreed that we might change this in the future, once we have all the use cases implemented.